concretizer: option to disable compiler mixing between runtime deps (#51135)
LLNL/spack: lib/spack/docs/configuring_compilers.rst

----------------

For more options on configuring Spack to mix different compilers for different languages, see :ref:`the toolchains configuration docs <toolchains>`.

To disable mixing (e.g. if you have multiple compilers defined, but want each concretized DAG to use one of them consistently), you can set:

.. code-block:: yaml

   concretizer:
     compiler_mixing: false

> This affects your root spec and everything it links to (transitively).

Reviewer: ```suggestion
This affects root specs and any (transitive) link or run dependencies.
```

Build-only dependencies are allowed to use different compilers (even when this is set).

> To enable mixing for specific packages, you can make that a whitelist:

Reviewer: ```suggestion
Some packages are difficult to build with high performance compilers, and it may be necessary to enable compiler mixing just for those packages.
To enable mixing for specific packages, specify an allow-list in the ``compiler_mixing`` config:
```

.. code-block:: yaml

   concretizer:
     compiler_mixing: ["openssl"]

> Note that this would not allow mixing for dependencies of ``openssl``.

Reviewer: ```suggestion
.. note::
  Adding ``openssl`` to the compiler mixing allow-list does not allow mixing for dependencies of ``openssl``.
```

----------------

For more options on configuring Spack to mix different compilers for different languages, see :ref:`the toolchains configuration docs <toolchains>`.

To disable mixing (e.g. if you have multiple compilers defined, but want each concretized DAG to use one of them consistently), you can set:

.. code-block:: yaml

   concretizer:
     compiler_mixing: false

This affects root specs and any (transitive) link or run dependencies.
Build-only dependencies are allowed to use different compilers (even when this is set).

.. note::

   Mixing detection is based on identifying compilers by name and version.

>    If you have two compilers that differ only in terms of variants, ``compiler_mixing:false`` would allow using both in a DAG.

Author: I think this addresses https://github.com/spack/spack/pull/51135#discussion_r2417336472

----------------

For more options on configuring Spack to mix different compilers for different languages, see :ref:`the toolchains configuration docs <toolchains>`.

To disable mixing (e.g. if you have multiple compilers defined, but want each concretized DAG to use one of them consistently), you can set:

.. code-block:: yaml

   concretizer:
     compiler_mixing: false

This affects root specs and any (transitive) link or run dependencies.
Build-only dependencies are allowed to use different compilers (even when this is set).

.. note::

   Mixing detection is based on identifying compilers by name and version.

>    If you have two compilers that differ only in terms of variants, ``compiler_mixing:false`` would allow using both in a DAG.

Reviewer 2: If we go with the implementation based on nodes, we can delete this:

```suggestion
```

Author: I tested node-based identity and it ran into https://github.com/spack/spack/pull/51135#discussion_r2437745875, so I think name/version-based identity will be needed.

EDIT: perhaps in https://github.com/spack/spack/pull/51135#discussion_r2445563338 you are suggesting a way to do this with node-identity - I'll check that.

Author: To confirm: this passes if I apply this and update the `count`. I'm set up a side meeting to resolve some questions I have about the logic, but something like this will land tomorrow in this PR and it should no longer rely on compiler name/version for identity.

Some packages are difficult to build with high performance compilers, and it may be necessary to enable compiler mixing just for those packages.
To enable mixing for specific packages, specify an allow-list in the ``compiler_mixing`` config:

.. code-block:: yaml

   concretizer:
     compiler_mixing: ["openssl"]

> .. note::

Reviewer 2: (nitpick, if you agree) We can probably put this outside of the note, just as a following sentence in the docs.
